### ISSUE TITLE
Adding NTP

### ISSUE THREAD

**Issue by michaelmargolis** [g0hjq/lcd_nixie_clock#2]
Many thanks for this Pico code.  If you would like to add NTP time sync using a Pico W, feel free to use the  code in the attached Zip file.

In addition to adding the included files supporting WiFi and NTP, there are minor changes needed to the Main and Display modules.

Code added to main.py to instantiate the wireless network as well as a time utility module for converting NTP time into micropython’s date and time formats. In this code, daylight savings settings are automatically  applied  if in the EU (inc UK) or North America.  (other regions could be added if needed)

The show_time tick handler method  in main.py has an added check to see if NTP sync is needed at startup and once per hour.

The only change to the display.py module is moving the display buffer to the top of the file so it can allocate memory at startup before other code fragments memory.  This is necessary on the Pico W because it has less memory than the non wifi version.

There are some enhancements and simplifications made possible using  WiFi and NTP that we can discuss if you are interested. For example:
 
	Day of week  and date displayed in rightmost LCD
	Web page for simple configuration setting

Again, thanks for your work making this possible.

[ntp_files.zip](https://github.com/g0hjq/lcd_nixie_clock/files/14232375/ntp_files.zip)

**Comment by g0hjq** [comment]
Hi Michael,

Thank you very much for getting in touch. I'm really pleased that other
people are finding the software useful.

I think I did originally try adding NTP support, but kept running out of
memory due to the extra ram needed by the WiFi adaptor. I really like your
idea of declaring the buffer before the include files. I've not seen that
before, and will definitely bear it in mind for future projects.

When I get the opportunity, I'll give your code a try on my own clock.
Would it be OK if I were to add it into my GitHub page, with a proper
reference to yourself of course? I might create a new branch to keep the
Pico and PicoW solutions separate.

I'd certainly be interested in any other enhancements you can come up with,
bearing in mind the limited memory - the font files take up quite a large
chunk of flash.

Gary.

**Comment by michaelmargolis** [comment]
Hi Gary, feel free to use any or all of that code.

A separate branch for the NTP version makes sense as some of the time setting code could be removed to allow more room for Pico W capabilities  such as web browser based configuring. Attached is a screen grab of a browser configuration screen that I am testing with the clock.

![config_page](https://github.com/g0hjq/lcd_nixie_clock/assets/579682/79638858-dc4b-41d0-9fbe-0186411d2c2a)

**Comment by michaelmargolis** [comment]
Here is the first cut  NTP time and date sync with Daylight savings support  along with a browser UI for settings.  The includes memory testing code that I will remove after more thorough testing. Please do test this yourself and let me know what needs fixing
[ntp_nixie_beta.zip](https://github.com/g0hjq/lcd_nixie_clock/files/14363601/ntp_nixie_beta.zip)

Due to the memory constraints of the Pico W running a web server, there were a number of changes to the implementation needed to get everything working.   Other than the settings workflow, the clock functionality has not been intentionally changed, but let me know if something does not work as needed.

**The functional changes to the settings comprises the following:**
•	Clock ‘fonts’ settings are now identified by name: nixie, 7 segment, dot matrix
•	The  colour of each clock font can be set to any 16 bit websafe colour
•	The LED’s colour can be set to  any 16 bit websafe colour or to the colour of the current clock font
                The brightness of the LEDs can be set to a value between 1 and 100 percent.	
•	A show_date toggle can be set to show the month and day on the rightmost digit (in hr//min mode). This also shows the alarm time if alarms are enabled. 
•	Hours offset from UTC can be set to adjust NTP time to local time.  
•	A DST option can be selected to: turn on or turn off DST, or to automatically set DST for EU/UK countries or North America.  Other DST regions could be added if needed.
•	The settings user interface is now through a web page, see example below
•	The supported button presses are:
o	 a short press on the mode button clears the currently active alarm and a long press (longer than two seconds) on the mode button will toggle the recurring alarm setting on or off. 
o	Short press on the right button to toggle the display of seconds on or off
o	Short press on the left button cycles through each of the fonts, each press selects another font.
o	Note these selections are not saved in the settings.json file

**Implementation changes made to enable the above functional additions:**
•	Changes to the settings dictionary are now made through the web interface. Existing code for button handling was removed. This enabled removing almost half of the lines of code in main.py. 
•	The intercommunication between the main code and the web code was made much simpler by structuring the clock and alarm code in classes.  
•	There is now a Clock class that encapsulates time display and interaction with the RTC. An Alarm class encapsulates checking and handling of the alarms .
•	A new button module is used to provide an event handling interface  for button presses
•	Display now takes a colour value along with font style when setting font. All colours in settings are in 16 bit web format, these are converted to the internal rgb565 in the display module
•	DS3231has an added Set_DateTime that accepts the python datetime tuple. Where appropriate, time references use standard python format and calls (i.e. localtime  to get the datetime tuple).
•	Leds.py has a method added to handle colour values in web format 

**New modules:**
•	Wifi.py and secrets.py  put your ssid and password in secrets.py, first connection my need a few retries
•	Webserver.py provides a UI for setting displayed in a browser.
•	Ntptime.py  returns utc time using the standard python datetime tuple
•	Time_utils contains code for syncing with the NTP code. It also has DST code that returns True if the current time is DST in the given region. The North America logic has not been tested.
•	Buttons.py provides an interface consistent with the polling interface of the other modules.

**Issues** 	
•	It is suggested that this version is maintained in a separate branch from the original version. Although some file names in this version are the same, they are not all compatible with the original code.
•	UK/EU is the only region tested for auto DST, NA should work but not tested
•	Browser may require refresh after sending settings in order to fully display all fields  after a change.
•	The RTC is synced with NTP  every 60 minutes so the  trim setting setting is probably not needed.   The value in settings.json is sent to the RTC but the web UI does not present this value.

  Web UI
![image](https://github.com/g0hjq/lcd_nixie_clock/assets/579682/751b47d8-7074-4542-b9b3-cc8ec4c3ce30)

**Comment by g0hjq** [comment]
Thank you ever so much Michael, that's absolutely brilliant. 

Do I have the latest version? I've given it a quick try and found the following:

1. Line 283 of main.py - I get a program error saying 2 parameters expected but only 1 supplied. I've commented out the lines below to get it working for now.
```       
if net.is_connected():
   if t_utils.check_sync():   <--- This fails
   # here is machine.RTC was set via NTP
      clock.set_rtc() # this sets the ds3231
```
2. On the web page, the image is not visible, just a place holder.
![image](https://github.com/g0hjq/lcd_nixie_clock/assets/37076748/49d39640-b933-4dc7-a6fd-97196142aa09)

3. When using the buttons to switch from 4-digit mode to 6-digit mode, the 3rd LCD (Where the colon was) is not being redrawn, resulting in the ten-minute digit being blank until it changes.

Your changes are quite substantial, so Instead of me making a branch for Wifi, would you prefer to take a fork into your own github repository? I can then add a link from my readme.md , and you one back to me.

**Comment by michaelmargolis** [comment]
Yes, its a versioning problem, the check_sync should be passed the method to sync the clock. As you suggested, best if I fork your repo repo and get everything consistent .  I will not be able to work on this until next week but please do report on any other issues.
The web image, if its in the images folder,  should appear after a while. The limited ram means that it is read in smallish chunks. Once read in, your browser should cache it.  Later I  may be able to increase the buffer size to reduce the time it takes to load the image.

**Comment by michaelmargolis** [comment]
Attached is the correct version of main.py  This should fix the call to check_sync.
I am looking into the 4 to 6 digit issue. 

Let me know what else needs attention.
[main.zip](https://github.com/g0hjq/lcd_nixie_clock/files/14365755/main.zip)

**Comment by michaelmargolis** [comment]
the fix for the missing digit when switching show_secs mode is to add a line that clears the digit cache

```
161            self.lcd.show_colon(2, digits[5]%2) # on every other second
162           self.digits_cache[2] = None  # Add this line to clear the digit cache <---
```

**Comment by michaelmargolis** [comment]
The NTP code is now online at:  https://github.com/michaelmargolis/lcd_ntp_nixie_clock
Please let me know if you have any additions or changes you would like for the functionality, code or documentation.
And again, thanks for doing the heavy lifting in getting this project going.

**Comment by g0hjq** [comment]
Hi Michael,

I've finally got round to installing your new NTP code on my clock. Thanks
for all your work. I really like your web server code, and the way you've
managed to squeeze everything into the limited space.

Just a few observations:

1. Your github repository doesn't contain a secrets.py file. It would be
useful if it had a one with the ssid and passwords blanked out.

2. If I'm in dot-matrix mode (and probably 7-segment mode) and change the
colour, the LCDs remain at the old colour until the digit changes. You
probably need to invalidate the digits cache on colour change.

3. I hit a very odd problem where the program sometimes crashed on
start-up. Investigation showed it was this line failing:

      t_utils =
time_utils.Time_utils(int(settings.get_setting("utc_offset")))

   What seems to be happening is that if I use the web interface to change
the display style, the utc_offset is being reset to empty string "".
   It looks as if the web browser is not always returning the expected
value "0". It only seems to go wrong on my android phone, and
   not the PC, and even then not always. Very strange!

Hope you have a nice bank holiday.
Gary.

**Comment by michaelmargolis** [comment]
Gary, thanks for the feedback.

Code posted to the repo should fix the issues you found: https://github.com/michaelmargolis/lcd_ntp_nixie_clock
The intermittant problem was due to the web server attempting to parse posted data before all data was sent by the browser.  I have revised the server code to better handle multiple packets.

Please let me know if you find any issues with the latest code.

Michael
